File: application/src/main/java/de/pbauerochse/worklogviewer/fx/SettingsViewController.java
package de.pbauerochse.worklogviewer.fx;

import de.pbauerochse.worklogviewer.WorklogViewer;
import de.pbauerochse.worklogviewer.connector.YouTrackConnectorLocator;
import de.pbauerochse.worklogviewer.connector.YouTrackVersion;
import de.pbauerochse.worklogviewer.fx.converter.WorkhoursStringConverter;
import de.pbauerochse.worklogviewer.fx.converter.YouTrackVersionStringConverter;
import de.pbauerochse.worklogviewer.settings.SettingsUtil;
import de.pbauerochse.worklogviewer.settings.SettingsViewModel;
import javafx.application.Platform;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.*;
import javafx.stage.Window;
import javafx.stage.WindowEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URL;
import java.util.ResourceBundle;
import java.util.stream.Stream;

/**
 * FX Controller for the settings dialog
 */
public class SettingsViewController implements Initializable {

    private static final Logger LOGGER = LoggerFactory.getLogger(SettingsViewController.class);

    @FXML
    private TextField youtrackUrlField;

    @FXML
    private ComboBox<YouTrackVersion> youtrackVersionField;

    @FXML
    private TextField youtrackUsernameField;

    @FXML
    private PasswordField youtrackPermanentTokenField;

    @FXML
    private ComboBox<Float> workhoursComboBox;

    @FXML
    private ComboBox<Theme> themeComboBox;

    @FXML
    private CheckBox showAllWorklogsCheckBox;

    @FXML
    private CheckBox showStatisticsCheckBox;

    @FXML
    private CheckBox loadDataAtStartupCheckBox;

    @FXML
    private CheckBox showDecimalsInExcel;

    @FXML
    private CheckBox enablePlugins;

    @FXML
    private Button saveSettingsButton;

    @FXML
    private Button cancelSettingsButton;

    @FXML
    private CheckBox mondayCollapseCheckbox;
    @FXML
    private CheckBox tuesdayCollapseCheckbox;
    @FXML
    private CheckBox wednesdayCollapseCheckbox;
    @FXML
    private CheckBox thursdayCollapseCheckbox;
    @FXML
    private CheckBox fridayCollapseCheckbox;
    @FXML
    private CheckBox saturdayCollapseCheckbox;
    @FXML
    private CheckBox sundayCollapseCheckbox;

    @FXML
    private CheckBox mondayHighlightCheckbox;
    @FXML
    private CheckBox tuesdayHighlightCheckbox;
    @FXML
    private CheckBox wednesdayHighlightCheckbox;
    @FXML
    private CheckBox thursdayHighlightCheckbox;
    @FXML
    private CheckBox fridayHighlightCheckbox;
    @FXML
    private CheckBox saturdayHighlightCheckbox;
    @FXML
    private CheckBox sundayHighlightCheckbox;

    private ResourceBundle resourceBundle;

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        LOGGER.debug("Initializing");
        this.resourceBundle = resources;

        SettingsViewModel viewModel = SettingsUtil.getSettingsViewModel();

        attachListeners(viewModel);
        initializeDefaultValues();
        bindInputElements(viewModel);
    }

    private void initializeDefaultValues() {
        // Workhours Combobox
        Stream.iterate(1f, n -> n <= 25f, n -> n + 0.25f).forEach(workhoursComboBox.getItems()::add);
        workhoursComboBox.setConverter(WorkhoursStringConverter.INSTANCE);

        // Version Combobox Values
        youtrackVersionField.getItems().addAll(YouTrackConnectorLocator.getSupportedVersions());
        youtrackVersionField.setConverter(YouTrackVersionStringConverter.INSTANCE);

        // Theme Combobox
        themeComboBox.getItems().addAll(Theme.values());
    }

    private void bindInputElements(SettingsViewModel viewModel) {
        youtrackUrlField.textProperty().bindBidirectional(viewModel.getYouTrackUrlProperty());
        youtrackVersionField.valueProperty().bindBidirectional(viewModel.getYouTrackVersionProperty());
        youtrackUsernameField.textProperty().bindBidirectional(viewModel.getYouTrackUsernameProperty());
        youtrackPermanentTokenField.textProperty().bindBidirectional(viewModel.getYouTrackPermanentTokenProperty());

        themeComboBox.valueProperty().bindBidirectional(viewModel.getThemeProperty());
        workhoursComboBox.valueProperty().bindBidirectional(viewModel.getWorkhoursProperty().asObject());
        showAllWorklogsCheckBox.selectedProperty().bindBidirectional(viewModel.getShowAllWorklogsProperty());
        showStatisticsCheckBox.selectedProperty().bindBidirectional(viewModel.getShowStatisticsProperty());
        loadDataAtStartupCheckBox.selectedProperty().bindBidirectional(viewModel.getLoadDataAtStartupProperty());
        showDecimalsInExcel.selectedProperty().bindBidirectional(viewModel.getShowDecimalsInExcelProperty());
        enablePlugins.selectedProperty().bindBidirectional(viewModel.getEnablePluginsProperty());

        mondayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateMondayProperty());
        tuesdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateTuesdayProperty());
        wednesdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateWednesdayProperty());
        thursdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateThursdayProperty());
        fridayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateFridayProperty());
        saturdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateSaturdayProperty());
        sundayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.getCollapseStateSundayProperty());

        mondayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateMondayProperty());
        tuesdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateTuesdayProperty());
        wednesdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateWednesdayProperty());
        thursdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateThursdayProperty());
        fridayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateFridayProperty());
        saturdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateSaturdayProperty());
        sundayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.getHighlightStateSundayProperty());
    }

    private void attachListeners(SettingsViewModel viewModel) {
        SimpleBooleanProperty hadMissingSettingsWhenOpened = new SimpleBooleanProperty(viewModel.getHasMissingConnectionSettings().get());

        // enable cancel button only when settings are valid or have been valid
        // when the form was shown, to allow the user cancel invalid inputs
        cancelSettingsButton.disableProperty().bind(hadMissingSettingsWhenOpened.and(viewModel.getHasMissingConnectionSettings()));

        cancelSettingsButton.setOnAction(event -> {
            LOGGER.debug("Cancel clicked");
            viewModel.discardChanges();
            closeSettingsDialogue();
        });

        saveSettingsButton.disableProperty().bind(viewModel.getHasMissingConnectionSettings());
        saveSettingsButton.setOnAction(event -> {
            LOGGER.debug("Save settings clicked");
            viewModel.saveChanges();
            closeSettingsDialogue();
        });
    }

    @FXML
    public void showSettingsHelp() {
        String helpUrl = resourceBundle.getString("view.settings.authentication.help_url");
        LOGGER.debug("Opening page {} in browser", helpUrl);
        Platform.runLater(() -> WorklogViewer.getInstance().getHostServices().showDocument(helpUrl));
    }

    private void closeSettingsDialogue() {
        LOGGER.debug("Closing settings dialogue");
        Window window = saveSettingsButton.getScene().getWindow();
        window.fireEvent(new WindowEvent(window, WindowEvent.WINDOW_CLOSE_REQUEST));
    }

}

File: application/src/main/java/de/pbauerochse/worklogviewer/fx/SettingsViewController.kt
package de.pbauerochse.worklogviewer.fx

import de.pbauerochse.worklogviewer.WorklogViewer
import de.pbauerochse.worklogviewer.connector.YouTrackConnectorLocator
import de.pbauerochse.worklogviewer.connector.YouTrackVersion
import de.pbauerochse.worklogviewer.fx.converter.WorkhoursStringConverter
import de.pbauerochse.worklogviewer.fx.converter.YouTrackVersionStringConverter
import de.pbauerochse.worklogviewer.fx.shortcutkeys.KeyCombinationAsStringBinding
import de.pbauerochse.worklogviewer.fx.shortcutkeys.KeyboardShortcutDefinition
import de.pbauerochse.worklogviewer.fx.shortcutkeys.RecordKeyboardShortcutListener
import de.pbauerochse.worklogviewer.settings.SettingsUtil
import de.pbauerochse.worklogviewer.settings.SettingsViewModel
import de.pbauerochse.worklogviewer.util.FormattingUtil.getFormatted
import javafx.application.Platform
import javafx.beans.property.SimpleBooleanProperty
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.geometry.Insets
import javafx.geometry.VPos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.GridPane
import javafx.scene.layout.RowConstraints
import javafx.stage.WindowEvent
import org.slf4j.LoggerFactory
import java.net.URL
import java.util.*
import java.util.stream.Stream

/**
 * FX Controller for the settings dialog
 */
class SettingsViewController : Initializable {

    @FXML
    private lateinit var youtrackUrlField: TextField
    @FXML
    private lateinit var youtrackVersionField: ComboBox<YouTrackVersion>
    @FXML
    private lateinit var youtrackUsernameField: TextField
    @FXML
    private lateinit var youtrackPermanentTokenField: PasswordField
    @FXML
    private lateinit var workhoursComboBox: ComboBox<Float>
    @FXML
    private lateinit var themeComboBox: ComboBox<Theme>
    @FXML
    private lateinit var showAllWorklogsCheckBox: CheckBox
    @FXML
    private lateinit var showStatisticsCheckBox: CheckBox
    @FXML
    private lateinit var loadDataAtStartupCheckBox: CheckBox
    @FXML
    private lateinit var showDecimalsInExcel: CheckBox
    @FXML
    private lateinit var enablePlugins: CheckBox
    @FXML
    private lateinit var saveSettingsButton: Button
    @FXML
    private lateinit var cancelSettingsButton: Button
    @FXML
    private lateinit var mondayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var tuesdayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var wednesdayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var thursdayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var fridayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var saturdayCollapseCheckbox: CheckBox
    @FXML
    private lateinit var sundayCollapseCheckbox: CheckBox

    @FXML
    private lateinit var mondayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var tuesdayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var wednesdayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var thursdayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var fridayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var saturdayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var sundayHighlightCheckbox: CheckBox
    @FXML
    private lateinit var keyboardShortcutGridPane: GridPane

    override fun initialize(location: URL, resources: ResourceBundle) {
        LOGGER.debug("Initializing")
        val viewModel = SettingsUtil.settingsViewModel

        attachListeners(viewModel)
        initializeDefaultValues()
        bindInputElements(viewModel)

        themeComboBox.sceneProperty().addListener { _, _, scene ->
            if (scene != null) {
                initializeKeyboardShortcutPanel(
                    scene,
                    listOf(
                        KeyboardShortcutDefinition(getFormatted("view.settings.shortkey.fetchworklogs"), viewModel.fetchWorklogsKeyboardCombination),
                        KeyboardShortcutDefinition(getFormatted("view.settings.shortkey.showissuesearch"), viewModel.showIssueSearchKeyboardCombination),
                        KeyboardShortcutDefinition(getFormatted("view.settings.shortkey.showsettings"), viewModel.showSettingsKeyboardCombination),
                        KeyboardShortcutDefinition(getFormatted("view.settings.shortkey.togglestatistics"), viewModel.toggleStatisticsKeyboardCombination),
                        KeyboardShortcutDefinition(getFormatted("view.settings.shortkey.exit"), viewModel.exitWorklogViewerKeyboardCombination)
                    )
                )
            }
        }
    }

    private fun initializeDefaultValues() {
        // Workhours Combobox
        Stream.iterate(1f, { n -> n <= 25f }, { n -> n!! + 0.25f }).forEach { workhoursComboBox.items.add(it) }
        workhoursComboBox.converter = WorkhoursStringConverter

        // Version Combobox Values
        youtrackVersionField.items.addAll(YouTrackConnectorLocator.getSupportedVersions())
        youtrackVersionField.converter = YouTrackVersionStringConverter

        // Theme Combobox
        themeComboBox.items.addAll(*Theme.values())
    }

    private fun bindInputElements(viewModel: SettingsViewModel) {
        youtrackUrlField.textProperty().bindBidirectional(viewModel.youTrackUrlProperty)
        youtrackVersionField.valueProperty().bindBidirectional(viewModel.youTrackVersionProperty)
        youtrackUsernameField.textProperty().bindBidirectional(viewModel.youTrackUsernameProperty)
        youtrackPermanentTokenField.textProperty().bindBidirectional(viewModel.youTrackPermanentTokenProperty)

        themeComboBox.valueProperty().bindBidirectional(viewModel.themeProperty)
        workhoursComboBox.valueProperty().bindBidirectional(viewModel.workhoursProperty.asObject())
        showAllWorklogsCheckBox.selectedProperty().bindBidirectional(viewModel.showAllWorklogsProperty)
        showStatisticsCheckBox.selectedProperty().bindBidirectional(viewModel.showStatisticsProperty)
        loadDataAtStartupCheckBox.selectedProperty().bindBidirectional(viewModel.loadDataAtStartupProperty)
        showDecimalsInExcel.selectedProperty().bindBidirectional(viewModel.showDecimalsInExcelProperty)
        enablePlugins.selectedProperty().bindBidirectional(viewModel.enablePluginsProperty)

        mondayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateMondayProperty)
        tuesdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateTuesdayProperty)
        wednesdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateWednesdayProperty)
        thursdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateThursdayProperty)
        fridayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateFridayProperty)
        saturdayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateSaturdayProperty)
        sundayCollapseCheckbox.selectedProperty().bindBidirectional(viewModel.collapseStateSundayProperty)

        mondayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateMondayProperty)
        tuesdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateTuesdayProperty)
        wednesdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateWednesdayProperty)
        thursdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateThursdayProperty)
        fridayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateFridayProperty)
        saturdayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateSaturdayProperty)
        sundayHighlightCheckbox.selectedProperty().bindBidirectional(viewModel.highlightStateSundayProperty)
    }

    private fun initializeKeyboardShortcutPanel(scene: Scene, shortcutDefinitions: List<KeyboardShortcutDefinition>) {
        val recordKeyboardShortcutListener = RecordKeyboardShortcutListener(scene)

        val cancelHintLabel = Label(getFormatted("view.settings.shortkey.cancelhint"))
        cancelHintLabel.styleClass.add("hint-label")
        cancelHintLabel.visibleProperty().bind(recordKeyboardShortcutListener.enabledProperty)

        shortcutDefinitions.forEachIndexed { index, definition ->
            val descriptionLabel = Label(definition.label)
            val anchorPane = AnchorPane()
            keyboardShortcutGridPane.addRow(index, descriptionLabel, anchorPane)

            GridPane.setColumnIndex(descriptionLabel, 0)
            GridPane.setColumnIndex(anchorPane, 1)
            GridPane.setValignment(anchorPane, VPos.CENTER)
            GridPane.setMargin(anchorPane, Insets(2.0, 0.0, 0.0, 0.0))

            val assignedShortcutLabel = Label()
            assignedShortcutLabel.textProperty().bind(KeyCombinationAsStringBinding(definition.property))

            val assignShortcutButton = Button(getFormatted("view.settings.shortkey.define"))
            assignShortcutButton.disableProperty().bind(recordKeyboardShortcutListener.enabledProperty)
            assignShortcutButton.onAction = EventHandler { recordKeyboardShortcutListener.listen(definition.property) }

            anchorPane.children.addAll(assignedShortcutLabel, assignShortcutButton)
            AnchorPane.setRightAnchor(assignShortcutButton, 0.0)

            keyboardShortcutGridPane.rowConstraints.add(RowConstraints(10.0, 30.0, 30.0))
        }

        keyboardShortcutGridPane.addRow(keyboardShortcutGridPane.rowCount, cancelHintLabel)
        GridPane.setColumnIndex(cancelHintLabel, 1)
        GridPane.setMargin(cancelHintLabel, Insets(4.0, 0.0, 0.0, 0.0))
    }

    private fun attachListeners(viewModel: SettingsViewModel) {
        val hadMissingSettingsWhenOpened = SimpleBooleanProperty(viewModel.hasMissingConnectionSettings.get())

        // enable cancel button only when settings are valid or have been valid
        // when the form was shown, to allow the user cancel invalid inputs
        cancelSettingsButton.disableProperty().bind(hadMissingSettingsWhenOpened.and(viewModel.hasMissingConnectionSettings))

        cancelSettingsButton.setOnAction {
            LOGGER.debug("Cancel clicked")
            viewModel.discardChanges()
            closeSettingsDialogue()
        }

        saveSettingsButton.disableProperty().bind(viewModel.hasMissingConnectionSettings)
        saveSettingsButton.setOnAction {
            LOGGER.debug("Save settings clicked")
            viewModel.saveChanges()
            closeSettingsDialogue()
        }
    }

    @FXML
    fun showSettingsHelp() {
        val helpUrl = getFormatted("view.settings.authentication.help_url")
        LOGGER.debug("Opening page {} in browser", helpUrl)
        Platform.runLater { WorklogViewer.instance.hostServices.showDocument(helpUrl) }
    }

    private fun closeSettingsDialogue() {
        LOGGER.debug("Closing settings dialogue")
        val window = saveSettingsButton.scene.window
        window.fireEvent(WindowEvent(window, WindowEvent.WINDOW_CLOSE_REQUEST))
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(SettingsViewController::class.java)
    }

}
